concept3.py

```diff
@@ -0,0 +1,108 @@
+import sys
+from prompt_toolkit import PromptSession
+import signal
+import asyncio
+
+
+class MentatInterrupt(Exception):
+    pass
+
+
+class MentatInterface:
+    # should be implemented by subclasses
+    def display(self, content, color=None, end="\n"):
+        pass
+
+    # should be implemented by subclasses
+    async def get_user_input(self, prompt=None, options=None):
+        pass
+
+    # convenience function
+    async def ask_yes_no(self, prompt=None):
+        return (await self.get_user_input(prompt, options=["y", "n"])).lower() == "y"
+
+
+class TerminalMentatInterface(MentatInterface):
+    """Used to communicate with the terminal"""
+
+    def __init__(self):
+        self.session = PromptSession()
+
+    def display(self, content, color=None, end="\n"):
+        print(content, end=end, flush=True)
+        global sigint
+        if sigint:
+            sigint = False
+            raise MentatInterrupt()
+
+    async def get_user_input(self, prompt=None, options=None):
+        while True:
+            if prompt is not None:
+                self.display(f"{prompt} ({'/'.join(options)})")
+
+            user_input =  await self.session.prompt_async()
+
+            if options is None:
+                return user_input
+            elif user_input.lower() in options:
+                return user_input
+            else:
+                self.display("Invalid response")
+
+
+async def run_mentat(interface):
+    for i in range(5):
+        print(i)
+        await asyncio.sleep(1)
```

Comment:
during this 5 second count, if you hit ctrl-c the signal handler does work - it'll print, and after this the global will be read and Mentat will exit